Fix REPL.fuzzyscore() function using Levenshtein distance and converting to a range from 0 to 1

PhoenixFlame101 opened the pull request:
Closes #49466

 - Implements `fuzzyscore()` using `levenshtein()`, and calculates the score by limiting it to a range from 0 to 1.
 - Works similar to `stringsimilarity()` as described [here](https://github.com/JuliaLang/julia/issues/49466#issuecomment-1525785061).
 
Here's a few examples:

    julia> REPL.fuzzyscore("abc", "xyz")
    0.0

    julia> REPL.fuzzyscore("abc", "abc")
    1.0

    julia> REPL.fuzzyscore("bupercalifragilisticexpialidocious", "supercalifragilisticexpialidocious")
    0.9705882352941176

    julia> REPL.fuzzyscore("IJilia", "MotionCaptureJointCalibration")
    0.1724137931034483

PhoenixFlame101 committed:
Fix REPL.fuzzyscore() function using Levenshtein distance

PhoenixFlame101 committed:
Add tests for new definition of REPL.fuzzyscore()

jakobnissen commented:
Thanks for the PR! There are some test failures that look legit. It also looks like `avgdistance` is not used elsewhere. Can it be deleted?

jakobnissen commented:
The test failure appears to be that the `printmatches` is used, when the user types a help prompt in the REPL, to check whether the query matches multiple symbols. For example:
```julia
help?> !
search: ! != !== sum! put! pop! map! get! any! all!
```

And this is done with `fuzzyscore(word, match) < 0 && break`, but of course this never happens anymore. So, you need a new heuristic to determine when something is so far away that it's not relevant for users.

PhoenixFlame101 committed:
Add heuristic for determining if two strings are too disimilar to match

 - Heuristic determines that two strings do not match if fuzzyscore is
   less than 60% (0.6)
 - Removed trailing whitespace in test/docview.jl

PhoenixFlame101 committed:
Remove avgdistance() from docview.jl

KristofferC commented:
Still looks like some CI errors:

```
REPL                                             (13) \|         failed at 2023-05-03T13:00:03.048
--
Test Failed at /cache/build/default-amdci4-7/julialang/julia-master/julia-c15110f193/share/julia/stdlib/v1.10/REPL/test/docview.jl:34
Expression: occursin("'mutable struct'", str)
Evaluated: occursin("'mutable struct'", "search:\n\nCouldn't find 'mutable s'\nPerhaps you meant ismutable\n")
```

tecosaur commented:
It's probably worth noting that levenshtein isn't great at transposition typos, e.g.
```julia
julia> 1 - REPL.levenshtein("hello", "helllo") / 6
0.8333333333333334

julia> 1 - REPL.levenshtein("hello", "ehllo") / 5
0.6
```

LilithHafner commented on stdlib/REPL/test/docview.jl:
```diff
@@ -46,14 +46,19 @@ end
+
+    @test REPL.fuzzyscore("abc", "xyz") == 0
+    @test REPL.fuzzyscore("abc", "abc") == 1
+    @test REPL.fuzzyscore("abc", "dcab") == 0.25
+    @test REPL.fuzzyscore("abc", "dab") ≈ 1 / 3
```
These tests seem fragile to me. I would prefer to see
```suggestion
    @test REPL.fuzzyscore("IJilia", "IJulia") >
          REPL.fuzzyscore("IJilia", "MotionCaptureJointCalibration")
    @test REPL.fuzzyscore("bupercalifragilisticexpialidocious", "supercalifragilisticexpialidocious") >
          REPL.fuzzyscore("parasid", "supercalifragilisticexpialidocious")
```
From @jakobnissen's examples in #49466

jakobnissen commented:
I'm going to close this in favor of #50412 . Nonetheless, thanks for giving it a go, @PhoenixFlame101 !
